ray-curator/ray_curator/stages/modules/filter.py

```diff
@@ -0,0 +1,500 @@
+    Then, determines whether to keep the document based on the criteria in the DocumentFilter.
+
+    The filter can be applied to any field in the dataset, and the score can be logged for later.
+    Also, the filter can be inverted such that "rejected" documents are kept.
+
+    Args:
+        filter_obj (DocumentFilter): The score function that takes in a document string and outputs a score for the document.
+        text_field (str): The field the documents will be read from.
+        score_field: The field to which the scores will be written. If None, scores will be immediately discarded after use.
+        score_type (Union[type, str]): The datatype of the score that will be made for each document.
+        id_field (str | None): The field to use as the document ID. Required for batch_size > 1.
+        invert (bool): If True, will keep all documents that are normally discarded.
+        processing_batch_size (int): The number of tasks to process in a batch.
+
+    """
+
+    filter_obj: DocumentFilter
+    text_field: str = "text"
+    score_field: str | None = None
+    score_type: type | str | None = None
+    id_field: str | None = None
+    invert: bool = False
+    processing_batch_size: int = 1
+
+    @property
+    def name(self) -> str:
+        return self.filter_obj.name
+
+    def inputs(self) -> tuple[list[str], list[str]]:
+        return ["data"], [self.text_field]
+
+    def outputs(self) -> tuple[list[str], list[str]]:
+        return ["data"], [self.score_field] if self.score_field else []
+
+    @property
+    def batch_size(self) -> int:
+        """Number of tasks to process in a batch."""
+        return self.processing_batch_size
+
+    def setup_on_node(self, _node_info: NodeInfo, _worker_metadata: WorkerMetadata) -> None:
+        if self.name in ["lang_id", "fasttext_quality_filter"]:
+            if not os.path.exists(self.filter_obj._model_path):
+                msg = f"Model file {self.filter_obj._model_path} not found"
+                raise FileNotFoundError(msg)
+        elif self.name in ["token_count"]:
+            # Use snapshot_download to download all files without loading the model into memory.
+            huggingface_hub.snapshot_download(
+                repo_id=self.filter_obj._hf_model_name,
+                token=self.filter_obj._hf_token,
+                local_files_only=False,  # Download if not cached
+                resume_download=True,  # Resume interrupted downloads
+            )
+
+    def setup(self, _: WorkerMetadata) -> None:
+        if self.name in ["lang_id", "fasttext_quality_filter"]:
+            self.filter_obj.model = fasttext.load_model(self.filter_obj._model_path)
+        elif self.name in ["token_count"] and self.filter_obj._tokenizer is None:
+            self.filter_obj._tokenizer = AutoTokenizer.from_pretrained(
+                self.filter_obj._hf_model_name,
+                local_files_only=True,  # Fail if not cached
+            )
```

Comment:
This seems to be common code relevant to `DocumentFilter` class, maybe move to a utility there and use it across Score / Filter / ScoreFilter instead of redoing it?

Reply:
Maybe you can do something like 
```python
def setup(self, _: WorkerMetadata) -> None:
    if isnstance(self.filter_obj, DocumentFilter):
        if hasattr(self.filter_obj, "load_model_fn"):
            self.filter_obj.load_model_fn() # this modifies self.filter_obj, therefore you don't need to modify state here
        if hasattr(self.filter_obj, "load_tokenizer_fn"):
            self.filter_obj.load_tokenzer_fn()
```

Similarly for setup_on_node
```python
def setup(self, _: WorkerMetadata) -> None:
    if isinstance(self.filter_obj, DocumentFilter);
          if hasattr(self.filter_obj, model_check_or_download()):
                 self.filter_obj, model_check_or_download()
```